### PR TITLE
Jacobian chunking in `VMC_SRt`

### PR DESCRIPTION
**Pull request by attila-i-szabo**
Some neural networks require much more memory for backpropagation, which makes it useful to allow chunking (with smaller chunks) when computing the Jacobian. (cf. #1347, #1590) 

This PR therefore adds Jacobian chunking to `VMC_SRt`. IIUC all this takes is to add an extra argument to `nkjax.jacobian` and copying some logic from `MCState` to set safe chunk sizes.

Outstanding questions:
* if we don't specify a chunk size explicitly, should `VMC_SRt` take the vstate's chunk size (like the QGT objects) or not chunk at all?
* what tests are needed?

### CODE REVIEW

**Comment by attila-i-szabo** [comment]
cc @riccardo-rende @llviteritti

**Comment by PhilipVinc** [comment]
What is wrong with #1590? It works well modulo a typo in `estate.expect_and_grad`. 
The only reason I have not merged the PR is because tests are missing.

(On a side note, I was thinking just now that maybe a better name would be `jac_chunk_size` instead of `grad_chunk_size`, but that is minor)

`SRt` should eventually get integrated into standard `SR` if we find a clean way to do it, so I'm a bit against doing this. I'd rather this be a general feature that benefits everyone. Moreover this is a bit confusing, because until now we keep the chunking in the variational state, while with this we'd have it in two different places.

I would be much happier if this was based of #1590, and `vstate.jac_chunk_size` was used as the chunk size if necessary.

**Comment by attila-i-szabo** [comment]
> What is wrong with #1590? 

It's not merged

> It works well modulo a typo in `estate.expect_and_grad`. The only reason I have not merged the PR is because tests are missing.
> 
> (On a side note, I was thinking just now that maybe a better name would be `jac_chunk_size` instead of `grad_chunk_size`, but that is minor)
> 
> I would be much happier if this was based of #1590, and `vstate.jac_chunk_size` was used as the chunk size if necessary.

You have to discuss with @chrisrothUT 

> `SRt` should eventually get integrated into standard `SR` if we find a clean way to do it, so I'm a bit against doing this. I'd rather this be a general feature that benefits everyone. 

That would be nice, but in the meanwhile it doesn't hurt to make `SRt` more usable. I'm also a little skeptical whether the two workflows can be merged nicely (I guess that's partly why it hasn't been done yet).

> Moreover this is a bit confusing, because until now we keep the chunking in the variational state, while with this we'd have it in two different places.

No, `QGTJacobian` also does its own chunking (i.e., calls `nkjax.jacobian` with a `chunk_size` argument, like my proposal here)

**Comment by PhilipVinc** [comment]
> No, QGTJacobian also does its own chunking (i.e., calls nkjax.jacobian with a chunk_size argument, like my proposal here)

Yes and no. `QGTJacobian` [defaults to the chunk size of the variational state](https://github.com/netket/netket/blob/c841005b3ab3fc09f419a8bbe5b12ef2064b6be6/netket/optimizer/qgt/qgt_jacobian_dense.py#L125) if it is there, even if it is possible to override it.

> It's not merged. Also, it only implements a different backprop chunk size for expect_and_* not for all backpropagation generally.

It does not, because it is not finished, but it introduces a standard way, consistent with what is there already to set the chunk size of backpropagations. It is trivial then to default to `jac_chunk_size` instead of `chunk_size` in QGT and similar things.
Moreover, we can easily have that `jac_chunk_size` defaults to `chunk_size` if it is unset.

`VMC_SRt` is supposed to be a stopgap until someone has the time to merge it back into `VMC` itself. 
I am not in favour of introducing new arguments to it that do not match into the model of `VMC` itself.
I am in favour of introducing a `jac_chunk_size` in the variational state rather than in `VMC_SRt`, following the idea of #1590, and using that in `VMC_SRt`.

**Comment by attila-i-szabo** [comment]
Ah okay, I thought you had some problem with doing chunking in general.

I don't have time to finish #1590, but it would be important to support chunking in `VMC_SRt` to some extent. How about we just chunk with `vstate.chunk_size` for now?

BTW, I think `grad_chunk_size` is better because we wouldn't always use it to compute a Jacobian. But that's beside the point of this PR.

**Comment by attila-i-szabo** [comment]
> `VMC_SRt` is supposed to be a stopgap until someone has the time to merge it back into `VMC` itself.

I don't have time for this either, but we should think about whether this is actually feasible/worth it.
The strategy of `VMC` is (1) compute energy gradients (2) apply some optional precoditioner on them. `SRt` just doesn't fit into this model. Of course you can always shoehorn two very different kinds of logic into one class, the question is if it's worth it.
One civilised alternative might be some kind of `PrecoditionedGradient` interface, which takes a vstate and spits out updates. But this too would take a lot of reshuffling...

**Comment by PhilipVinc** [comment]
This is now implemented in the new VMC_SR
